subsys: fs: Support file open flags to fs and POSIX API

de-nordic opened the pull request:
**{DNM]** Due to required update to mcumgr, that is reviewed here: https://github.com/apache/mynewt-mcumgr/pull/90
and requires update of https://github.com/zephyrproject-rtos/mcumgr

**Anyone** that is able to review https://github.com/apache/mynewt-mcumgr/pull/90 please do so!

This series of commits adds support for file open flags (CREATE, APPEND, etc.) and fs_open_ex function to support them.
POSIX API was updated accordingly.

- [x] FAT FS tests
- [x] LIttleFS test
- [x] Posix tests

RFC Issue: https://github.com/zephyrproject-rtos/zephyr/issues/26833
Resolves: https://github.com/zephyrproject-rtos/zephyr/issues/16638

~~Requires changes to mcumgr: https://github.com/apache/mynewt-mcumgr/pull/90~~
Requires changes to mcumgr, for the regression to pass, brought by snapshot update: https://github.com/zephyrproject-rtos/mcumgr/pull/26

de-nordic committed:
subsys/fs: Add open flags parameter to fs_open

The fs_open has been extended with support for open flags.
Currently supported flags are:
  FS_O_READ -- open for read
  FS_O_WRITE -- open for write
  FS_O_CREATE -- create file if it does not exist
  FS_O_APPEND -- move to the end of file before each write

The FAT FS and LittleFS front-ends within the Zephyr has also been
modified to utilize the flags.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/subsys/fs: Add fs_open flags tests to fat_fs_api

The addition of support for open flags, within fs_open, requires
addition of new test cases to FAT FS back-end tests.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/subsys/fs: Add common fs_open_flags test procedure

Common procedure, to use with fs back-ends, have been added for testing
various fs_open flags combinations.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/subsys/fs: Add fs_open flags tests to LittleFS

The addition of support for open flags, within fs_open, requires
addition of new test cases.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
lib/posix: Add support for open flags to open(...)

The fs_open flags has been changed to accept open flags, which requires
changes to open(...) to support the new flags.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

jimparis commented:
POSIX is an API specification, so "adding to posix API" does not make sense, because then it's not posix anymore.  I think we should instead fix `open()` (by not unconditionally adding `O_CREAT`) and not introduce `open2()` at all.

jimparis reviewed (changes requested):
I think we should fix open() (by not unconditionally adding O_CREAT) and not introduce open2() at all.

jimparis commented on lib/posix/fs.c:
```diff
@@ -54,17 +57,38 @@ static inline void posix_fs_free_obj(struct posix_fs_desc *ptr)
 	ptr->used = false;
 }
 
+static int posix_mode_to_zephyr(int mf)
+{
+	int mode = 0;
+
+	mode = (mf & O_ACCMODE) + 1;
+
+	if (mf & O_CREAT) {
+		if (!(mode & FO_WRITE)) {
+			errno = EINVAL;
+		} else {
+			mode |= FO_CREATE;
+		}
+	}
+	return mode;
+}
+
```
What about `O_APPEND` (which is defined in newlib, but missing from zephyr's minimal libc).  This function makes numeric assumptions about the values of O_RDWR etc; it would be more robust to specifically check flags, e.g.:
```
if (mf & O_RDWR)
    mode = FO_RDWR;
else if ...
```

de-nordic replied:
Done

jimparis commented on subsys/fs/fat_fs.c:
```diff
@@ -63,7 +63,35 @@ static int translate_error(int error)
+		fat_mode |= FA_OPEN_APPEND;
+		flags ^= FO_APPEND;
+	}
+
```
I believe the intention was to warn here about unsupported flags, as in the littlefs version of the same code?
If not, there's no reason to clear bits in `flags`.

de-nordic replied:
Will fix.

jimparis commented on tests/subsys/fs/littlefs/src/test_lfs_basic.c:
```diff
@@ -473,6 +475,198 @@ static int check_large(void)
+		      0,
+		      "verify open existing failed");
+
+	/* NOTE: The Zephyr fs_open_ex flags allow opening existing files
+	 * with READ/WRITE access, as it happens in fs_open_ex above, but
+	 * the LittleFS assumes READ/WRITE access by default if no access
+	 * is selected which same as requesting READ ONLY and WRITE ONLY
+	 * access at the same time (bug?).
+	 * From Zephyr API point of view below tests should fail but they
+	 * will succeed, which is expected (?) behaviour of LittleFS;
+	 * that is why this test is left here commented out.
```
If I'm understanding you right, you are saying that littlefs requires flags to be either `LFS_O_RDONLY`, `LFS_O_WRONLY`, or `LFS_O_RDWR`, and passing a flag value of 0 is treated the same as `LFS_O_RDWR`.  If that's true, shouldn't it be the job of `lfs_flags_from_zephyr` to never pass a flag value of 0 to lfs?

de-nordic replied:
`lfs_flags_from_zephyr ` is supposed to translate flags to obtain, from fs backend, specific behavior. I guess that the READONLY would be closest to expected.

de-nordic commented:
> 
> 
> POSIX is an API specification, so "adding to posix API" does not make sense, because then it's not posix anymore. I think we should instead fix `open()` (by not unconditionally adding `O_CREAT`) and not introduce `open2()` at all.

I am not sure I can do that immediately because someone may be using current implementation the way it is. If it just gets changed then the code will stop working. Maybe there should be some Kconfig added that would enable compatibility with the original behavior for some time.

de-nordic committed:
tests/posix/fs: Add tests for open flags

Add new tests that will utilize open mode flags with the open
function.

A few test functions which assumed O_CREAT behavior are updated
to specify it explicitly.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

jimparis commented on lib/posix/Kconfig:
```diff
@@ -91,6 +91,21 @@ config POSIX_FS
+
+config POSIX_FS_LEGACY_OPEN
+	bool "Enable legacy open(...) behaviour"
+	default y if FILE_SYSTEM_LEGACY_OPEN
```
This line does nothing because `FILE_SYSTEM_LEGACY_OPEN` isn't true here

jimparis reviewed (dismissed):
I dislike `FILE_SYSTEM_LEGACY_OPEN`.  I suggest using some macro tricks like the ones described at https://stackoverflow.com/questions/3046889/optional-parameters-with-c-macros, which will let users continue to pass either 2 or 3 arguments to `fs_open`, so that both existing code and new code can coexist and work correctly.  For example, this test appears to do the trick:
```c
#include <stdio.h>

#define FO_READ   0x01
#define FO_WRITE  0x02
#define FO_RDWR   (FO_READ | FO_WRITE)
#define FO_CREATE 0x10

void real_fs_open(void *p, const char *path, int flags) {
        printf("open(p=%p, path=\"%s\", flags=0x%02x)\n", p, path, flags);
}

void legacy_fs_open(void *p, const char *path) {
        return real_fs_open(p, path, FO_CREATE | FO_RDWR);
}

/* With 0-2 arguments, fs_open(...) resolves to legacy_fs_open(...)
   With 3   arguments, fs_open(...) resolves to real_fs_open(...)
   More arguments are not handled.
*/
#define GET_ARG4(arg1, arg2, arg3, arg4, ...) arg4
#define OPEN_CHOOSE(...) GET_ARG4(                                      \
                __VA_ARGS__,  real_fs_open,                             \
                legacy_fs_open, legacy_fs_open, legacy_fs_open)
#define fs_open(...) OPEN_CHOOSE(__VA_ARGS__)(__VA_ARGS__)

/* Use example: */
main()
{
        char p;

        fs_open(&p, "/no-flags");
        fs_open(&p, "/with-flags", FO_READ);

        return 0;
}
```
Output:
```
$ gcc -Wall -pedantic -o test test.c
$ ./test
open(p=0x7fff85ee440f, path="/no-flags", flags=0x13)
open(p=0x7fff85ee440f, path="/with-flags", flags=0x01)
```
As for the POSIX side, I would just fix it and not support the old broken behavior at all.  If it's still made configurable, I would strongly suggest making it default to the correct open() API; otherwise it doesn't make sense to call it POSIX at all, and I am certain that anyone attempting to port POSIX code to Zephyr would welcome the fix.

de-nordic commented:
> 
> I dislike `FILE_SYSTEM_LEGACY_OPEN`. I suggest using some macro tricks (...)

OK I can do that. I just hoped to deprecate the old function signature with time and leave the 3 parameter thing only.

> 
> As for the POSIX side, I would just fix it and not support the old broken behavior at all (...)

I do not know how many users have been using this in the form it is. The "legacy" behavior is supposed to be deprecated after some time. The programmers should be given time to adjust code. If we will keep on breaking their applications without giving time to adjust to changes, they will be reluctant to update Zephyr within their projects.

jimparis commented:
> > I dislike `FILE_SYSTEM_LEGACY_OPEN`. I suggest using some macro tricks (...)
> 
> OK I can do that. I just hoped to deprecate the old function signature with time and leave the 3 parameter thing only.

That sounds good.  You could use the macro and mark the old signature with `__attribute__((deprecated))` so GCC will warn if it gets used.

> > As for the POSIX side, I would just fix it and not support the old broken behavior at all (...)
> 
> I do not know how many users have been using this in the form it is. The "legacy" behavior is supposed to be deprecated after some time. The programmers should be given time to adjust code. If we will keep on breaking their applications without giving time to adjust to changes, they will be reluctant to update Zephyr within their projects.

I don't think it's a big deal either way, but I would still learn towards fixing it because:
* the POSIX API is clear
* the open() function signature isn't changing; i.e. users have always had to pass flags to it
* Previously, the flags were just ignored
* Now, the flags are interpreted correctly

So basically the only case where someone would get bitten by the change is if they were previously passing flags that didn't match the behavior they actually wanted -- ie. if they were passing "O_RDONLY" and then expecting to be able to write the file.

I would suggest naming the configuration option `POSIX_FS_LEGACY_OPEN_IGNORE_FLAGS` with a description like: 
> With this option set, the POSIX `open()` call will ignore the user-supplied flags, and act as if `O_CREAT|O_RDWR` were passed.  For backwards compatibility with Zephyr < A.B.C.  This option will default to `n` starting in Zephyr X.Y.Z.

Unfortunately, because `open()` already took flags, it's harder to help the user out with compile-time warnings about the upcoming deprecation.  Maybe use `__builtin_constant_p` in a macro to issue a warning if the passed flags are something other than `O_CREAT|O_RDWR`, with the legacy option?  It wouldn't warn for non-constant flags, but should still cover some use cases.

de-nordic commented:
> I don't think it's a big deal either way, but I would still learn towards fixing it because:
> 
>     * the POSIX API is clear
> 
>     * the open() function signature isn't changing; i.e. users have always had to pass flags to it
> 
>     * Previously, the flags were just ignored
> 
>     * Now, the flags are interpreted correctly

OK, I agree with the argumentation. The Kconfig will has been removed.

pfalcon commented on tests/posix/fs/CMakeLists.txt:
```diff
@@ -4,5 +4,11 @@ cmake_minimum_required(VERSION 3.13.1)
-FILE(GLOB app_sources src/*.c)
-target_sources(app PRIVATE ${app_sources})
+target_sources(app PRIVATE src/main.c src/test_fat_mount.c)
+if (TEST_ONLY_OPEN_FLAGS)
```
Why the need for this ifdeffery? I'd say the new testcase source file should be just added to the existing test application, without any conditions.

de-nordic replied:
OK, will separate testcases.

de-nordic replied:
Merged instead of separating but "ifdeffery" has been removed.

de-nordic replied:
Done, removed.

pfalcon commented:
> As for the POSIX side, I would just fix it and not support the old broken behavior at all. 

+1. As a POSIX subsys maintainer, I wouldn't like to deal with supporting buggy behavior, given that we have many more bugs to fix and many more features to implement, and should concentrate on that.

pfalcon commented on tests/posix/fs/src/test_fs_dir.c:
```diff
@@ -27,7 +27,8 @@ static int test_mkdir(void)
 	}
 
-	res = open(TEST_DIR_FILE, O_RDWR);
+	res = open(TEST_DIR_FILE, O_CREAT | O_RDWR);
```
Please extend the commit message of this commit with something like "A few test functions which assumed O_CREAT behavior are updated to specify it explicitly."

de-nordic replied:
Done.

pfalcon commented on tests/posix/fs/src/test_fs_file.c:
```diff
@@ -16,12 +16,13 @@ static int test_file_open(void)
 	int res;
 
-	res = open(TEST_FILE, O_RDWR);
+	res = open(TEST_FILE, O_RDWR | O_CREAT);
```
A mior nit, but can you please sort O_RDWR to the end - they will be consistent with a change above, and a usual code style (main op, then RD/WR flags).

de-nordic replied:
OK

de-nordic replied:
Done.

pfalcon commented on lib/libc/minimal/include/fcntl.h:
```diff
@@ -8,13 +8,14 @@
 #define F_SETFL 4
 
-int open(const char *name, int flags);
+int open(const char *name, int flags, ...);
```
Please add a note to the commit message that you change open() signature. Note that while this change is correct (and I'm ok with it being piggy-backed here), nothing in the rest of your patch relies or uses this change. So, if you follow a "don't make unrelated/unnecessary changes" strictly, you could as well drop this change (and leave it for later, for dedicated commit, and detailed commit message).

de-nordic replied:
OK, will move it to separate commit with proper comment. The change has been done, because some of headers come from SDK and there the `open` is declared with ellipsis, which caused conflict with Zephyr definition of function.

pfalcon replied:
> OK, will move it to separate commit with proper comment. 

That's how I'd do it, so appreciated. Otherwise, I agree it's the right change to do.

de-nordic replied:
Done.

pfalcon reviewed (dismissed):
Thanks for the patch. I concentrate on review the POSIX API part.

First of all, thanks for already dealing away with "open2()" function.

I still think there're relatively minor, but still changes to do. With details in inline comments, general nature of my suggestion is:

1. Please make the commit messages more detailed, describing what and why being done, because some pretty important changes go unexplained.
2. I wouldn't like to see arbitrary restrictions to be encoded in the POSIX API layer. For example, there's nothing wrong with creating a file not available for writing (to the caller). If a particular FS implementation has restrictions, then please validate flags on its level (if it's not yet done).
3. I'm not sure why new POSIX tests are added in the "ifdef" manner.They rather should be unconditionally integrate in the existing testsuite.

pfalcon commented:
And minor optimization note:

~~~
#define O_APPEND   0x0400
#define FO_APPEND	0x20
~~~

If you chose `FO_*` to match `O_*` values, then a smart modern compiler could optimize away a majority of code in posix_mode_to_zephyr().

de-nordic commented on lib/posix/fs.c:
```diff
@@ -54,17 +57,56 @@ static inline void posix_fs_free_obj(struct posix_fs_desc *ptr)
+		break;
+	}
+
+	if ((mode & FO_CREATE || mode & FO_APPEND) && !(mode & FO_WRITE)) {
```
OK. Actually I will take the suggestion down to Zephyr fs API  and change the fs_open function to be able to create file with no write access.

de-nordic replied:
Done.

pabigot commented on include/fs/fs.h:
```diff
@@ -160,6 +160,30 @@ struct fs_file_system_t {
+#define FO_MODE_BITS	0x03
+#endif
+
+#ifndef FO_READ
```
All these `#ifndef`s are weird: why would somebody define `FO_READ` before this header, and if they did wouldn't it be a problem?  That's a motivation for using an enum for the constants.

IMO these bits should be defined as `FS_MODE_READ` etc, preserving the existing `fs_` prefix for the public API in this header.  It could be `FS_OPEN_READ` if there's no possibility of using these for other roles (such as permissions).

de-nordic replied:
Ifdefs removed. I am not sure about `FS_` prefix. I have selected `FO_` for the file open, and this is similar to posix where open flags start with `O_`

pabigot replied:
API 2020-07-07 Per discussion prefer `FS_O_foo` to be namespaced in filesystem API but match POSIX `O_foo` flag names.

de-nordic replied:
Done.

pabigot commented on include/fs/fs.h:
```diff
@@ -170,20 +194,35 @@ struct fs_file_system_t {
+ *	   -ERRNO  other errno code, deending of file system backend
  */
-int fs_open(struct fs_file_t *zfp, const char *file_name);
+int fs_open(struct fs_file_t *zfp, const char *file_name, int flags);
```
I don't particularly like seeing signed types used to hold bitfields.  I guess existing practice in legacy APIs like POSIX might justify it.

de-nordic replied:
Done

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,13 +181,50 @@ static void release_file_data(struct fs_file_t *fp)
+
+	if (zflags & FO_CREATE) {
+		flags |= LFS_O_CREAT;
+		zflags ^= FO_CREATE;
```
What's with `^=`?  `&= ~` would be less startling (standard).

de-nordic replied:
Done.

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,13 +181,50 @@ static void release_file_data(struct fs_file_t *fp)
 	struct fs_littlefs *fs = fp->mp->fs_data;
 	struct lfs *lfs = &fs->lfs;
-	int flags = LFS_O_CREAT | LFS_O_RDWR;
+	int flags;
```
Ideally this (and `ret`) would be initialized in the variable definition, just like `lfs` is.

de-nordic replied:
OK

pabigot commented on tests/subsys/fs/fat_fs_api/src/test_fat_file.c:
```diff
@@ -38,7 +38,411 @@ static int test_file_open(void)
+{
+	int res;
+	struct fs_file_t fp;
+	char something[] = "Something";
```
This should be `const char* something = ` unless you intend to modify the string content (or `const char something[] = ` if you need to do `sizeof()`).

pabigot commented on tests/subsys/fs/fat_fs_api/src/test_fat_file.c:
```diff
@@ -38,7 +38,411 @@ static int test_file_open(void)
+
+	TC_PRINT("* Attempt to open non-existing file (expect fail)\n");
+	/* Should fail */
+	res = fs_open(&fp, FATFS_MNTP"/missing", 0);
```
IMO it's easier to read these if there's whitespace between the macro token and the string.  Though if this is the style used in this file, I guess we live with it (or change the existing uses in a separate whitespace-only commit).

de-nordic replied:
I am reworking tests into more readable form now, so do not bother looking at them for now.

de-nordic replied:
Done.

pabigot reviewed (changes requested):
I've added this to the API project as the changes to the FS API must be discussed.

My primary change requested is not using `FO_` as a prefix for definitions associated with the file system API, and some concerns about how those constants are defined and passed.

de-nordic commented:
> 
> 
> I've added this to the API project as the changes to the FS API must be discussed.
> 
> My primary change requested is not using `FO_` as a prefix for definitions associated with the file system API, and some concerns about how those constants are defined and passed.

OK, should I be present on some meetings or the discussion will be passed down here?

pabigot commented:
> > I've added this to the API project as the changes to the FS API must be discussed.
> 
> OK, should I be present on some meetings or the discussion will be passed down here?

It'd be good to be present, but you don't have to be.  Conclusions (if any) will be documented in a comment here.

de-nordic committed:
posix: Extend open() function signature with ellipsis

The commit changes signature of open function from:
  int open(const char *name, int flags)
to
  int open(const char *name, int flags, ...)

Currently existing two argument invocations should not require any
rework.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

pabigot commented:
API 2020-07-07 No objections.  @MaureenHelm asked whether there was a reason for not having these flags in the original API, if anybody has knowledge of that.

Next step is to follow [the steps required](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) to make a change to a stable API.

nvlsianpu reviewed (commented):
Just a few nips.

nvlsianpu commented:
@de-nordic I was initially thinking about refine added test case using `ztest_test_skip()` for skip not-relevant test primitives- but found it hard to be used here.

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,12 +181,50 @@ static void release_file_data(struct fs_file_t *fp)
-	int flags = LFS_O_CREAT | LFS_O_RDWR;
-	int ret;
+	int flags = 0;
+	int ret = 0;
```
Yes, what I was asking for (if this was in response to a request I remembered making) is:
```
int flags = fs_flags_from_zephyr(zflags);
int ret = k_mem_slab_alloc(&file_data_pool, &fp->filep, K_NO_WAIT);

if (ret != 0) {
...
```

de-nordic replied:
OK, will fix.

carlescufi commented:
@jimparis @pfalcon can you please take another look?

pfalcon commented on include/fs/fs.h:
```diff
@@ -160,6 +161,17 @@ struct fs_file_system_t {
 					struct fs_statvfs *stat);
 };
 
+#define FO_READ		0x01
```
> align with values below

I may imagine that's just a rendering artifact due to use of tabs. @de-nordic, can you confirm? One obvious way to avoid this confusion is to avoid use of tabs. Zephyr mandates use of tabs only for indentation (i.e., at the beginning of line), and the less we use tabs beyond that, the better.

de-nordic replied:
OK. Will fix.

pfalcon commented on include/fs/fs.h:
```diff
@@ -170,20 +182,32 @@ struct fs_file_system_t {
-
 /**
- * @brief File open
+ * @brief File open extended
```
I'm not sure adding word "extended" is useful here. (If this is extended, that implies somewhere there's "barebones" one, but there's no such. Indeed, having flags fro open is a "baseline". The fact that we didn't have them before is, well, a bug.)

de-nordic replied:
Will fix. This is leftover from having two fs_open functions, the documentation for the one supporting the flags have been left, but description has not been fixed.

de-nordic replied:
Done.

pfalcon commented on include/fs/fs.h:
```diff
@@ -170,20 +182,32 @@ struct fs_file_system_t {
+ * @retval -EINVAL when bad combination of flags is given
+ *	   -EINVAL when bad file name is given
+ *	   -NOENT  when file path is not possible (bad mount point)
+ *	   -ERRNO  other errno code, deending of file system backend
```
> -ERRNO  other errno code, deending of file system backend

That's ambiguous/confusing (and with a few typos). Suggestion: "other negative error code, depending on file system backend".

pabigot replied:
Also the way Doxygen works I believe `@retval` is required before each return value.  There's no implicit continuation onto subsequent lines.

de-nordic replied:
Will fix.

de-nordic replied:
Done.

pfalcon commented on include/fs/fs.h:
```diff
@@ -170,20 +182,32 @@ struct fs_file_system_t {
  */
-int fs_open(struct fs_file_t *zfp, const char *file_name);
+int fs_open(struct fs_file_t *zfp, const char *file_name, unsigned int flags);
+#define fs_open(f, ...) _fs_open(f, __VA_ARGS__, FO_CREATE | FO_RDWR)
```
Ah, so there's actually a magic to make older 2-arg fs_open() to work with the new 3-arg fs_open()? A bit too much magic to me, I hope somebody vetted that.

de-nordic replied:
Yes, I have not found a way to resolve this with C syntax only. Maybe someone does know about some extension into gcc and clang that would fix this. There is less magic here than in `OPEN_CHOOSE`, which is more advance usage of cpp than the two cpp lines used here.

pfalcon replied:
> Yes, I have not found a way to resolve this with C syntax only. Maybe ...

I wouldn't worry too much about it, if it works, it's ok. We should treat that as as a deprecated transitional feature, to be removed eventually.

de-nordic replied:
It has been decided to drop backward compatibility and note need for changes, in existing code, in release notes.

pfalcon commented:
Commit message: "posix: Extend open() function signature with ellipsis"

> Currently existing tow argument invocations should not require any
rework.

Typo: "tow", should be "two".

pfalcon commented:
> @pfalcon can you please take another look?

POSIX side of things looks good, thanks @de-nordic ! But looking thru other commits, I saw minor typos/things worth fixing. So, I just dismissed my older -1, to not hold it back, but don't give explicit +1 for now.

de-nordic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,12 +181,50 @@ static void release_file_data(struct fs_file_t *fp)
 	struct lfs *lfs = &fs->lfs;
-	int flags = LFS_O_CREAT | LFS_O_RDWR;
-	int ret;
+	int flags = 0;
```
OK

nvlsianpu reviewed (approved):
API enhancement accepted in advance to typo fixes.

pabigot commented on include/fs/fs.h:
```diff
@@ -170,20 +182,31 @@ struct fs_file_system_t {
+ * @param flags The mode flags, if not provided default to FO_CREATE | FO_RDWR
+ *
+ * @p flags can be empty, or combination of one or more of following flags:
+ *   FO_READ    open for read
```
Correct these flag names.

de-nordic replied:
Done

pabigot commented on subsys/fs/fs.c:
```diff
@@ -75,7 +76,7 @@ static int fs_get_mnt_point(struct fs_mount_t **mnt_pntp,
 
 /* File operations */
-int fs_open(struct fs_file_t *zfp, const char *file_name)
+int fs_open(struct fs_file_t *zfp, const char *file_name, unsigned int flags)
```
The third parameter should be of type:

    typedef uint8_t fs_mode_t;

not a bare unsigned int.

pfalcon replied:
> The third parameter should be of type:
typedef uint8_t fs_mode_t;
not a bare unsigned int.

Perhaps, but then it should be typedef to the very unsigned int, not an adhoc type like uint8_t. When passing params to funcs, native word sized params should be preferred, unless there's a specific need to do otherwise.

pabigot replied:
In POSIX the value passed through this argument is documented to be of type `mode_t`.

Since the FS API is not a POSIX API [coding guidelines Dir 4.6](https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html) suggest that it should be a standard sized type.

I don't particularly care what it's typedef'd as, but it needs to be a typedef'd type.

de-nordic replied:
OK, do you want it to be `uint8_t`? If ABI passes it through registers it will be of register size anyway (which is usually int for CPU) or passed via stack that is usually aligned to size of CPU register. Why to restrict it to `uint8_t`

de-nordic replied:
WIll make it `uint8_t`, the ",member of structure" argument by @pabigot is ok.

de-nordic replied:
Done.

pabigot commented on include/fs/fs.h:
```diff
@@ -160,6 +161,17 @@ struct fs_file_system_t {
+#define FS_O_MODE_MASK  0x03
+
+#define FS_O_CREATE     0x10
+#define FS_O_APPEND     0x20
```
I would be a little careful about adding `O_APPEND` support.  In POSIX and littlefs this means atomically seek to the end of the file before each write.  In FatFs it appears to mean "seek to end of file when opened".  If the desired semantics is the former then there is no equivalent FatFs flag and the FatFs wrapper would have to be modified after each seek.  If the desired semantics is a one-time move to the end this flag doesn't do that in littlefs so there is no equivalent flag and the littlefs wrapper would have to be modified to move when opened.

I don't see a test for the precise semantics (create a file with content X, then open it with append, write A, seek to the start, write B, and verify that content is XAB).  I suggest doing this, because it may be that FatFS does the same thing as littlefs (though the code suggests it doesn't.)

de-nordic replied:
Added the missing test scenario.

de-nordic replied:
I think that the desired outcome should be that writing to file opened for append should append data at the end of file. The test I have provided is very simple one, it just checks if file position changes to expected size after appending already existing file.

pabigot replied:
I think you've missed my point: that is **not** the semantics the flag will have on POSIX or littlefs (as documented).

Assume an application where a file is opened with `FS_O_APPEND`, some material is written, then the application seeks to the start, reads a little, then writes something.  We cannot have a situation where the file content is different based on whether the file system is littlefs or FatFs.

de-nordic replied:
OK, I understand now.

carlescufi replied:
API meeting 14th July:
- Adopt the POSIX understanding of `O_APPEND`
- Use the native LittleFS APPEND flag
- Modify  the FatFS shim so that it seeks to the end before every write when a file is opened with `O_APPEND`

de-nordic replied:
Will fix.

de-nordic replied:
Done.

de-nordic replied:
Done

pabigot commented on tests/subsys/fs/littlefs/src/test_lfs_basic.c:
```diff
@@ -80,8 +82,7 @@ static int create_write_hello(const struct fs_mount_t *mp)
 	zassert_equal(fs_open(&file,
-			      testfs_path_init(&path, mp,
-					       HELLO,
+			      testfs_path_init(&path, mp, HELLO,
```
Please avoid gratuitous reformatting until we have a tool that does it.  The line separation here was intentional.

Also: there is no change in this file other that the reformatting and introduction of two macros that aren't used.

de-nordic replied:
Removed

pabigot commented on tests/subsys/fs/littlefs/src/testfs_util.c:
```diff
@@ -235,7 +235,7 @@ int testfs_build(struct testfs_path *root,
 				     testfs_path_extend(root,
 							cp->name,
-							TESTFS_PATH_END));
+							TESTFS_PATH_END)),
```
This should remain a semicolon.

de-nordic committed:
doc: relnotes: Add API change note for fs_open

Update release notes API change section with information on change to
fs_open parameter list.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

pabigot commented on include/fs/fs.h:
```diff
@@ -170,20 +185,32 @@ struct fs_file_system_t {
-int fs_open(struct fs_file_t *zfp, const char *file_name);
+int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags);
+#define fs_open(f, ...) _fs_open(f, __VA_ARGS__, FS_O_CREATE | FS_O_RDWR)
+#define _fs_open(f, fp, flags, ...) fs_open(f, fp, flags)
```
This is a neat trick, but it fails to diagnose this misuse:
```
fs_open(fsp, "test", 3, 4);
```

de-nordic replied:
You are right here, because `...` in `_fs_open` discards extra parameters.

galak commented on include/fs/fs.h:
```diff
@@ -170,20 +185,32 @@ struct fs_file_system_t {
  */
-int fs_open(struct fs_file_t *zfp, const char *file_name);
+int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags);
+#define fs_open(f, ...) _fs_open(f, __VA_ARGS__, FS_O_CREATE | FS_O_RDWR)
```
Need to add a comment here about what is going on with the #define and VA_ARGS

de-nordic replied:
thanks for the comment, but this will be removed (https://github.com/zephyrproject-rtos/zephyr/pull/26305#discussion_r454484395)

carlescufi commented on include/fs/fs.h:
```diff
@@ -170,20 +185,32 @@ struct fs_file_system_t {
  */
-int fs_open(struct fs_file_t *zfp, const char *file_name);
+int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags);
+#define fs_open(f, ...) _fs_open(f, __VA_ARGS__, FS_O_CREATE | FS_O_RDWR)
```
API Meeting 14th July:
- Remove the macros
- Change the release notes entry so that users know that they need to adapt their code.

de-nordic replied:
Done.

de-nordic commented:
NOTE: There have been west.yml update commit added on the top of all commits, that contains update to mcumgr:
https://github.com/apache/mynewt-mcumgr/pull/90

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -143,10 +168,18 @@ static ssize_t fatfs_read(struct fs_file_t *zfp, void *ptr, size_t size)
 	unsigned int bw;
+	off_t pos = f_size((FIL *)zfp->filep);
+
+	if (((FIL *)zfp->filep)->flag & FA_SEEKEND) {
+		res = f_lseek(zfp->filep, pos);
+	}
```
Comment needed.

de-nordic replied:
Done

de-nordic committed:
gui: Add open flags argument to fs_open invocation

Now fs_open supports open flags and the lvgl_fs code requires update.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
native_posix: Add flags argument to fs_open invocations

Now fs_open supports open flags and the fuse requires update.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented on subsys/fs/fuse_fs_access.c:
```diff
@@ -345,7 +345,7 @@ static int fuse_fs_access_truncate(const char *path, off_t size)
 	static struct fs_file_t file;
 
-	err = fs_open(&file, path);
+	err = fs_open(&file, path, FS_O_CREATE | FS_O_RDWR);
```
Done

de-nordic commented on lib/gui/lvgl/lvgl_fs.c:
```diff
@@ -58,7 +58,7 @@ static lv_fs_res_t lvgl_fs_open(struct _lv_fs_drv_t *drv, void *file,
 	path--;
 
-	err = fs_open((struct fs_file_t *)file, path);
+	err = fs_open((struct fs_file_t *)file, path, FS_O_CREATE | FS_O_RDWR);
```
Done

pabigot commented on include/fs/fs.h:
```diff
@@ -170,20 +185,29 @@ struct fs_file_system_t {
+ *   FS_O_WRITE   open for write
+ *   FS_O_RDWR    open for read/write (<tt>FS_O_READ | FS_O_WRITE</tt>)
+ *   FS_O_CREATE  create file if it does not exist
+ *   FS_O_APPEND  open file at the end
```
This is incorrect.  POSIX semantics is "move to end before each write".  The file position is not moved to the end on open.

Please also fix in commit message.

de-nordic replied:
Done

pabigot commented on subsys/fs/fat_fs.c:
```diff
@@ -63,7 +65,20 @@ static int translate_error(int error)
+	fat_mode |= (flags & FS_O_READ) ? FA_READ : 0;
+	fat_mode |= (flags & FS_O_WRITE) ? FA_WRITE : 0;
+	fat_mode |= (flags & FS_O_CREATE) ? FA_OPEN_ALWAYS : 0;
+	fat_mode |= (flags & FS_O_APPEND) ? FA_SEEKEND : 0;
```
This should be named `FAX_APPEND`.  `FA_SEEKEND` is Fat-FS only and means "seek to end of file on open".  That's not something Zephyr supports, so we should use a dedicated name, with a prefix that doesn't conflict with what FatFS uses for its supported modes.

de-nordic replied:
Unfortunately "the hack" with using flag from FAT FS would not work, so I have decided to modify `fs_file_t` structure to contain copy of the flags and now all fs_ functions have access to original flags and can use them for different purposes.

pabigot commented on subsys/fs/fat_fs.c:
```diff
@@ -143,10 +158,23 @@ static ssize_t fatfs_read(struct fs_file_t *zfp, void *ptr, size_t size)
+	 * to always write at the end of file, so set file position
+	 * at the end before each write if FA_SEEKEND has been selected.
+	 */
+	if (((FIL *)zfp->filep)->flag & FA_SEEKEND) {
```
Update to `FAX_APPEND` as noted above.

de-nordic replied:
Done as explained in https://github.com/zephyrproject-rtos/zephyr/pull/26305#discussion_r457908605

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,14 +181,49 @@ static void release_file_data(struct fs_file_t *fp)
+		zflags &= ~FS_O_APPEND;
+	}
+
+	if (zflags != 0) {
```
I don't think it's worth emitting a warning for something that causes no harm.  This will only be a problem if/when we add more flags and don't update this module.  It would be reasonable to just get rid of this, and so also all the actions that clear the individual bits of `zflags` (which is also a little weird).

Supporting this perspective: you didn't check for extra flags in the fat_fs translation.  So please remove it here.

de-nordic replied:
Done.

pabigot commented on tests/subsys/fs/common/test_fs_open_flags.c:
```diff
@@ -0,0 +1,313 @@
+
+	ZBEGIN("Attempt append existing with no R/W access");
+	ZMKEMPTY(&ts);
+	ZOPEN(&ts, FS_O_APPEND,  0);
```
Add:

    ZCHKPOS(&ts, 0);

Also after all other opens that include `FS_O_APPEND`.

de-nordic replied:
Done.

pabigot commented on tests/subsys/fs/common/test_fs_open_flags.c:
```diff
@@ -0,0 +1,313 @@
+/* ZEQ decides whether test completed successfully and prints appropriate
+ * information.
+ */
+static void ZEQ(int ret, int expected)
```
Use lower case for all these function names: all-caps belongs to macros only.

The trick of `#DEFINE FOO(a) FOO(a,x)` is confusing because the "same" function appears to have two signatures; using the correct style for these names will eliminate that confusion.

pabigot commented on tests/subsys/fs/common/test_fs_open_flags.c:
```diff
@@ -146,6 +146,10 @@ void test_fs_open_flags(void)
 	ZOPEN(&ts, FS_O_WRITE, -ENOENT);
 	ZOPEN(&ts, FS_O_READ, -ENOENT);
 	ZOPEN(&ts, FS_O_RDWR, -ENOENT);
+	/* This is valid case for API, but FAT driver will always create,
```
Comment is out of date (at least by referencing `FO_`.  It's also very unclear: it suggests that FatFS is different that other file systems here; if it was, it wouldn't return `-ENOENT` so the test would fail, right?

Clarification is needed.  Best alternative is this comment is removed because it's no longer necessary.

pabigot replied:
Ah; I see the name at least gets fixed in patch 9.  Please move the fix to where the comment is introduced (or remove the comment there).

de-nordic replied:
Will fix. The commits in series should not fix each other, so there is mistake.

de-nordic replied:
Fixed.

pabigot commented on tests/subsys/fs/fat_fs_api/src/test_fat_file.c:
```diff
@@ -38,7 +38,7 @@ static int test_file_open(void)
 		TC_PRINT("File now exists %s\n", TEST_FILE);
 	}
 
-	return res;
```
Are these changes necessary?  I agree this source file is wrong in not explictly returning the `TC_PASS` it looks for, but if this needs to be done it should be in a separate cleanup commit, and all tests should be updated to do this (and the one that's not `static` made so).

de-nordic replied:
Changes removed.  Will fix in other PR.

pabigot commented on tests/subsys/fs/littlefs/CMakeLists.txt:
```diff
@@ -5,4 +5,11 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 
 FILE(GLOB app_sources src/*.c)
-target_sources(app PRIVATE ${app_sources})
+# LittleFS opens files for RW by default if no flags given and crashes,
```
Ick.  OK, I'll have to look into that.

pabigot commented on lib/posix/fs.c:
```diff
@@ -63,8 +95,13 @@ int open(const char *name, int flags, ...)
 {
 	int rc, fd;
 	struct posix_fs_desc *ptr = NULL;
+	int zmode;
```
Where possible initialize a variable when it's declared: make this:

    int zmode = posix_mode_to_zephyr(flags)

Unless @pfalcon objects.

de-nordic replied:
Done.

pabigot commented on doc/releases/release-notes-2.4.rst:
```diff
@@ -46,6 +46,13 @@ API Changes
+  All custom file system front-ends require change to the implementation
+  of ``open`` callback to accept the new parameter.
+  To maintain original behaviour within user code, two argument invocations
+  should be converted to accept third parameter of the value  ``FS_O_CREATE | FS_O_RDWR``.
```
"...should be converted to ~accept~*pass a* third ~parameter of the value~*argument*  ``FS_O_CREATE | FS_O_RDWR``."

de-nordic replied:
Done.

pabigot commented on include/fs/fs.h:
```diff
@@ -170,20 +182,29 @@ struct fs_file_system_t {
  *
- * Opens an existing file or create a new one and associates
- * a stream with it.
+ * Opens or crates, if does not exist file, depending on mode provided
```
s/crates/creates/

de-nordic replied:
Done

pabigot commented on lib/posix/fs.c:
```diff
@@ -54,17 +57,43 @@ static inline void posix_fs_free_obj(struct posix_fs_desc *ptr)
 
+static int posix_mode_to_zephyr(int mf)
+{
+	int mode = mf & O_CREAT ? FS_O_CREATE : 0;
```
Please use parentheses:

    int mode = (mf & O_CREAT) ? FS_O_CREATE : 0;

everywhere this idiom is used, to clarify intent and disambiguate from:

    int mode = mf & (O_CREAT ? FS_O_CREATE : 0);

which would be wrong.   It's not strictly necessary, but it is more clear, and it was done that one in one change already.

de-nordic replied:
Done

pabigot commented:
DNM pending merge of  mynewt-mcumgr pull in west.

pabigot reviewed (changes requested):
One style change, one spelling fix.

pabigot commented on subsys/fs/littlefs_fs.c:
```diff
@@ -181,14 +181,29 @@ static void release_file_data(struct fs_file_t *fp)
-static int littlefs_open(struct fs_file_t *fp, const char *path)
+static int lfs_flags_from_zephyr(unsigned int zflags)
+{
+	int flags = zflags & FS_O_CREATE ? LFS_O_CREAT : 0;
```
Missed this one.

de-nordic replied:
Fixed

de-nordic commented:
@pabigot Thanks!

de-nordic committed:
west.yml: Snapshot update of mcumgr from upstream

This commit updates mcumgr with latest snapshot from the upstream.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

carlescufi commented on west.yml:
```diff
@@ -92,7 +92,7 @@ manifest:
       path: bootloader/mcuboot
     - name: mcumgr
-      revision: 898a5a7f5224be8345e58eca3b9a84389ed61d15
+      revision: pull/26/head
```
Use cfe5eb98a9493017448846fd1a44a9340bd0a22f
